### DISCUSSION TITLE
[API Gateway] External X-Request-Id overridden by Envoy

### DISCUSSION THREAD

**Discussion by flower-of-the-bridges** [mia-platform/community#93] [Q&A] [answered]
Hi,

I'm using the API Gateway with Envoy.
If I try to perform a http request to an endpoint exposed by API Gateways's Envoy with the following header:

```
`X-Request-Id: prova`
```

I receive the following response header:

```
X-Request-Id: e8e65231-ae5a-4d58-8a92-1798a0de7a2b 
```

It seems that Envoy overrides the header with a UUID generated when a request is received.

On [envoy's documentation](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/observability/tracing) is stated that:

> Envoy will generate UUIDs when needed and populate the [x-request-id](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#config-http-conn-man-headers-x-request-id) HTTP header. Applications can forward the x-request-id header for unified logging as well as tracing. The behavior can be configured on a per [HTTP connection manager](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#envoy-v3-api-field-extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-request-id-extension) basis using an extension.

This behaviour can be modified in the [HttpConnectionManager extension](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#http-connection-manager-proto) by setting the parameter `preserve_external_request_id`.

How can we achieve the same result in the API Gateway advanced section?

**Comment by mattia-fumo** [comment, marked as answer]
Hi Giovanni, this is a perfect use case for the [listener patch extension](https://docs.mia-platform.eu/docs/development_suite/api-console/advanced-section/api-gateway-envoy/extensions#patching-arbitrary-listeners-properties).
Something like the following snippet will do the trick:
```yaml
- listener_name: LISTENER
  'filter_chains.0.filters.0.typed_config.preserve_external_request_id': true
```
PS. Remember to replace `LISTENER` with the actual listener name.

**Comment by davidebianchi** [comment]
Hi @giovf96, be careful to use this configuration! If the API Gateway is exposed on internet, it's a bad practice to accept a request id header instead of generating it by the gateway. It could expose you to possible attacks.

**Comment by flower-of-the-bridges** [comment]
Hi @mattia-fumo, thanks for your answer, the trick works!

@davidebianchi thanks for your suggestion: yes, the API is exposed on internet, but with all endpoints having user group permissions: can this avoid attacs? And what are possible scenarios where an attack can be performed?

I think also, If we're giving this "power" to users to edit listeners configurations, to tell what are possible parameters that could lead to expose unsecure endpoints

**Comment by mattia-fumo** [comment]
Nice, would you mind marking the question as answered? Thank you in advance
